main/src/ca/uwaterloo/flix/util/Validation.scala

```diff
@@ -149,34 +178,174 @@ object Validation {
     if (xs.isEmpty)
       return Validation.SuccessNil
 
-    // Two mutable arrays to hold the intermediate results.
-    val successValues = mutable.ArrayBuffer.empty[S]
-    val failureStream = mutable.ArrayBuffer.empty[LazyList[E]]
+    // Three mutable arrays to hold the intermediate results.
+    val successValues = mutable.ArrayBuffer.empty[(S, Int)]
+    val successFailureStream = mutable.ArrayBuffer.empty[(S, LazyList[E], Int)]
+    val failureStream = mutable.ArrayBuffer.empty[(LazyList[E], Int)]
 
     // Apply f to each element and collect the results.
-    for (x <- xs) {
+    for ((x, i) <- xs.zipWithIndex) {
       f(x) match {
-        case Success(v) => successValues += v
-        case Failure(e) => failureStream += e
+        case Success(v) => successValues += ((v, i))
+        case SuccessWithFailures(v, e) => successFailureStream += ((v, e, i))
+        case Failure(e) => failureStream += ((e, i))
       }
     }
 
     // Check whether we were successful or not.
-    if (failureStream.isEmpty) {
-      Success(successValues.toList)
+    if (!failureStream.isEmpty) {
+      val combined = combineFailureStreams(failureStream, successFailureStream)
+      Failure(combined)
+
+    } else if (!successFailureStream.isEmpty) {
+
+      val failures = successFailureStream.map {
+        case (_, e, _) => e
+      }.foldLeft(LazyList.empty[E])(_ #::: _)
+
+      val combined = combineSuccessStreams(successValues, successFailureStream)
+      SuccessWithFailures(combined, failures)
+
     } else {
-      Failure(failureStream.foldLeft(LazyList.empty[E])(_ #::: _))
+      val values = successValues.map {
+        case (v, _) => v
+      }.toList
+
+      Success(values)
     }
   }
 
+  /**
+    * Returns the combined stream of `failureStream` and `successFailureStream`.
+    *
+    * Helper function for `fastTraverse`.
+    */
+  private def combineFailureStreams[E, S, T](failureStream: ArrayBuffer[(LazyList[E], Int)], successFailureStream: ArrayBuffer[(S, LazyList[E], Int)]) = {
```

Comment:
This seems a bit crazy to me...

Reply:
Agreed, but how should we combine the streams if we lose the order of the errors and values?

Reply:
My comment was like: "I want a pony". Does not mean I can get it. I was just wondering if there is a better solution without offering any insight on how to do it.

Reply:
Can't we just keep the two failure streams and also a flag that says whether the error is fatal?

Reply:
> Can't we just keep the two failure streams and also a flag that says whether the error is fatal?

Can you sketch?

Reply:
```scala
    var isFatal = false
    // Apply f to each element and collect the results.
    for (x <- xs) {
      f(x) match {
        case Success(v) => successValues += v
        case SuccessWithFailures(v, e) =>
            successValues += v
            failureStream += e
        case Failure(e) => 
            failureStream += e
            isFatal = true
      }
    }


    // Check whether we were successful or not.
    if (isFatal) {
        /Failure(failureStream)
    } else if (!failureStream.isEmpty) {
      Success(successValues.toList)
    } else {
        SuccessWithFailures(successValues, failureStream)
    }
}
```